`java.lang.IndexOutOfBoundsException: Index -1 out of bounds for length 0`

### Plugin Modernizer version

plugin modernizer 999999-SNAPSHOT (2025-01-25T14:24:40Z)

### What Operating System are you using

Debian / WSL3

### Reproduction steps

`java -jar plugin-modernizer-cli/target/jenkins-plugin-modernizer-999999-SNAPSHOT.jar run --plugins valid-network --recipe UpgradeToRecommendCoreVersion`

### Expected Results

No error.

### Results
```
2025-01-25T14:36:09.723Z [INFO] [Thread=StreamPumper-systemOut] - i.j.t.p.core.impl.MavenInvoker # [INFO] Project is using Jenkins bom at version 3
2025-01-25T14:36:09.873Z [INFO] [Thread=StreamPumper-systemOut] - i.j.t.p.core.impl.MavenInvoker # [ERROR] The recipe produced an error. Please report this to the recipe author.
2025-01-25T14:36:09.878Z [INFO] [Thread=StreamPumper-systemOut] - i.j.t.p.core.impl.MavenInvoker # [INFO] ------------------------------------------------------------------------
2025-01-25T14:36:09.880Z [INFO] [Thread=StreamPumper-systemOut] - i.j.t.p.core.impl.MavenInvoker # [INFO] BUILD FAILURE
2025-01-25T14:36:09.881Z [INFO] [Thread=StreamPumper-systemOut] - i.j.t.p.core.impl.MavenInvoker # [INFO] ------------------------------------------------------------------------
2025-01-25T14:36:09.882Z [INFO] [Thread=StreamPumper-systemOut] - i.j.t.p.core.impl.MavenInvoker # [INFO] Total time:  22.180 s
2025-01-25T14:36:09.883Z [INFO] [Thread=StreamPumper-systemOut] - i.j.t.p.core.impl.MavenInvoker # [INFO] Finished at: 2025-01-25T15:36:09+01:00
2025-01-25T14:36:09.884Z [INFO] [Thread=StreamPumper-systemOut] - i.j.t.p.core.impl.MavenInvoker # [INFO] ------------------------------------------------------------------------
2025-01-25T14:36:09.885Z [INFO] [Thread=StreamPumper-systemOut] - i.j.t.p.core.impl.MavenInvoker # [ERROR] Failed to execute goal org.openrewrite.maven:rewrite-maven-plugin:6.0.5:run (default-cli) on project valid-network: Execution default-cli of goal org.openrewrite.maven:rewrite-maven-plugin:6.0.5:run failed: Error while visiting pom.xml: java.lang.IndexOutOfBoundsException: Index -1 out of bounds for length 0
2025-01-25T14:36:09.886Z [INFO] [Thread=StreamPumper-systemOut] - i.j.t.p.core.impl.MavenInvoker # [ERROR]   java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64)
2025-01-25T14:36:09.887Z [INFO] [Thread=StreamPumper-systemOut] - i.j.t.p.core.impl.MavenInvoker # [ERROR]   java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70)
2025-01-25T14:36:09.887Z [INFO] [Thread=StreamPumper-systemOut] - i.j.t.p.core.impl.MavenInvoker # [ERROR]   java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:266)
2025-01-25T14:36:09.888Z [INFO] [Thread=StreamPumper-systemOut] - i.j.t.p.core.impl.MavenInvoker # [ERROR]   java.base/java.util.Objects.checkIndex(Objects.java:361)
2025-01-25T14:36:09.889Z [INFO] [Thread=StreamPumper-systemOut] - i.j.t.p.core.impl.MavenInvoker # [ERROR]   java.base/java.util.ArrayList.get(ArrayList.java:427)
2025-01-25T14:36:09.890Z [INFO] [Thread=StreamPumper-systemOut] - i.j.t.p.core.impl.MavenInvoker # [ERROR]   io.jenkins.tools.pluginmodernizer.core.visitors.UpdateBomVersionVisitor.getLatestBomVersion(UpdateBomVersionVisitor.java:166)
2025-01-25T14:36:09.891Z [INFO] [Thread=StreamPumper-systemOut] - i.j.t.p.core.impl.MavenInvoker # [ERROR]   io.jenkins.tools.pluginmodernizer.core.visitors.UpdateBomVersionVisitor.getLatestBomVersion(UpdateBomVersionVisitor.java:109)
2025-01-25T14:36:09.892Z [INFO] [Thread=StreamPumper-systemOut] - i.j.t.p.core.impl.MavenInvoker # [ERROR]   io.jenkins.tools.pluginmodernizer.core.visitors.UpdateBomVersionVisitor.visitDocument(UpdateBomVersionVisitor.java:85)
2025-01-25T14:36:09.893Z [INFO] [Thread=StreamPumper-systemOut] - i.j.t.p.core.impl.MavenInvoker # [ERROR]   io.jenkins.tools.pluginmodernizer.core.visitors.UpdateBomVersionVisitor.visitDocument(UpdateBomVersionVisitor.java:25)
2025-01-25T14:36:09.894Z [INFO] [Thread=StreamPumper-systemOut] - i.j.t.p.core.impl.MavenInvoker # [ERROR]   org.openrewrite.xml.tree.Xml$Document.acceptXml(Xml.java:149)
2025-01-25T14:36:09.895Z [INFO] [Thread=StreamPumper-systemOut] - i.j.t.p.core.impl.MavenInvoker # [ERROR]   org.openrewrite.xml.tree.Xml.accept(Xml.java:53)
2025-01-25T14:36:09.896Z [INFO] [Thread=StreamPumper-systemOut] - i.j.t.p.core.impl.MavenInvoker # [ERROR]   org.openrewrite.TreeVisitor.visit(TreeVisitor.java:250)
2025-01-25T14:36:09.897Z [INFO] [Thread=StreamPumper-systemOut] - i.j.t.p.core.impl.MavenInvoker # [ERROR]   org.openrewrite.TreeVisitor.visit(TreeVisitor.java:157)
2025-01-25T14:36:09.898Z [INFO] [Thread=StreamPumper-systemOut] - i.j.t.p.core.impl.MavenInvoker # [ERROR]   org.openrewrite.Preconditions$Check.visit(Preconditions.java:175)
2025-01-25T14:36:09.899Z [INFO] [Thread=StreamPumper-systemOut] - i.j.t.p.core.impl.MavenInvoker # [ERROR]   org.openrewrite.Preconditions$Check.visit(Preconditions.java:145)
2025-01-25T14:36:09.899Z [INFO] [Thread=StreamPumper-systemOut] - i.j.t.p.core.impl.MavenInvoker # [ERROR]   org.openrewrite.scheduling.RecipeRunCycle.lambda$editSources$6(RecipeRunCycle.java:186)
2025-01-25T14:36:09.900Z [INFO] [Thread=StreamPumper-systemOut] - i.j.t.p.core.impl.MavenInvoker # [ERROR]   ...
2025-01-25T14:36:09.901Z [INFO] [Thread=StreamPumper-systemOut] - i.j.t.p.core.impl.MavenInvoker # [ERROR] -> [Help 1]
2025-01-25T14:36:09.902Z [INFO] [Thread=StreamPumper-systemOut] - i.j.t.p.core.impl.MavenInvoker # [ERROR]
2025-01-25T14:36:09.904Z [INFO] [Thread=StreamPumper-systemOut] - i.j.t.p.core.impl.MavenInvoker # [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
2025-01-25T14:36:09.905Z [INFO] [Thread=StreamPumper-systemOut] - i.j.t.p.core.impl.MavenInvoker # [ERROR] Re-run Maven using the -X switch to enable full debug logging.
2025-01-25T14:36:09.906Z [INFO] [Thread=StreamPumper-systemOut] - i.j.t.p.core.impl.MavenInvoker # [ERROR]
2025-01-25T14:36:09.907Z [INFO] [Thread=StreamPumper-systemOut] - i.j.t.p.core.impl.MavenInvoker # [ERROR] For more information about the errors and possible solutions, please read the following articles:
2025-01-25T14:36:09.909Z [INFO] [Thread=StreamPumper-systemOut] - i.j.t.p.core.impl.MavenInvoker # [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException
2025-01-25T14:36:09.959Z [ERROR] [Thread=main] - i.j.t.p.core.impl.MavenInvoker # Build fail with code: 1
```

--- a/plugin-modernizer-core/src/main/java/io/jenkins/tools/pluginmodernizer/core/config/RecipesConsts.java
+++ b/plugin-modernizer-core/src/main/java/io/jenkins/tools/pluginmodernizer/core/config/RecipesConsts.java
@@ -14,8 +14,9 @@ public final class RecipesConsts {
 
     public static final String PLUGIN_POM_GROUP_ID = "org.jenkins-ci.plugins";
     public static final String PLUGINS_BOM_GROUP_ID = "io.jenkins.tools.bom";
-    public static final String ANALYSIS_POM_GROUP_ID = "org.jvnet.hudson.plugins";
     public static final String ANALYSIS_POM_ARTIFACT_ID = "analysis-pom";
     public static final String VERSION_METADATA_PATTERN = "\\.v[a-f0-9_]+";
+    public static final String OLD_BOM_VERSION_PATTERN =
+            ""; // Old release line like 2.150.x, no need to filter by pattern
     public static final String INCREMENTAL_REPO_ID = "incrementals";
 }

--- a/plugin-modernizer-core/src/main/java/io/jenkins/tools/pluginmodernizer/core/visitors/UpdateBomVersionVisitor.java
+++ b/plugin-modernizer-core/src/main/java/io/jenkins/tools/pluginmodernizer/core/visitors/UpdateBomVersionVisitor.java
@@ -41,6 +41,12 @@ public class UpdateBomVersionVisitor extends MavenIsoVisitor<ExecutionContext> {
             new LatestRelease(RecipesConsts.VERSION_METADATA_PATTERN);
 
     /**
+     * Old bom version comparator that where not using JEP-229
+     */
+    private final transient LatestRelease oldBomReleaseComparator =
+            new LatestRelease(RecipesConsts.OLD_BOM_VERSION_PATTERN);
+
+    /**
      * Contructor
      */
     public UpdateBomVersionVisitor(MavenMetadataFailures metadataFailures) {
@@ -154,16 +160,27 @@ public class UpdateBomVersionVisitor extends MavenIsoVisitor<ExecutionContext> {
 
         // Keep track of version found
         List<String> versions = new ArrayList<>();
+        List<String> oldVersions = new ArrayList<>();
         for (String v : mavenMetadata.getVersioning().getVersions()) {
             if (latestBomReleaseComparator.isValid(currentVersion, v)) {
                 versions.add(v);
+            }
+            if (oldBomReleaseComparator.isValid(currentVersion, v)) {
+                oldVersions.add(v);
             }
         }
 
         // Take latest version available. Allow to downgrade from incrementals to release
         if (!Semver.isVersion(currentVersion) && !versions.isEmpty() || (!versions.contains(currentVersion))) {
             versions.sort(latestBomReleaseComparator);
-            return versions.get(versions.size() - 1);
+            oldVersions.sort(oldBomReleaseComparator);
+            if (!versions.isEmpty()) {
+                return versions.get(versions.size() - 1);
+            }
+            if (!oldVersions.isEmpty()) {
+                return oldVersions.get(oldVersions.size() - 1);
+            }
+            return currentVersion;
         } else {
             return latestBomReleaseComparator.upgrade(currentVersion, versions).orElse(null);
         }

--- a/plugin-modernizer-core/src/test/java/io/jenkins/tools/pluginmodernizer/core/recipes/DeclarativeRecipesTest.java
+++ b/plugin-modernizer-core/src/test/java/io/jenkins/tools/pluginmodernizer/core/recipes/DeclarativeRecipesTest.java
@@ -321,6 +321,101 @@ public class DeclarativeRecipesTest implements RewriteTest {
     }
 
     @Test
+    void testUpgradeOldBomVersionFormat() {
+        rewriteRun(
+                spec -> spec.recipeFromResource(
+                        "/META-INF/rewrite/recipes.yml", "io.jenkins.tools.pluginmodernizer.UpgradeBomVersion"),
+                // language=xml
+                pomXml(
+                        """
+                        <?xml version="1.0" encoding="UTF-8"?>
+                        <project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
+                          <modelVersion>4.0.0</modelVersion>
+                          <parent>
+                            <groupId>org.jenkins-ci.plugins</groupId>
+                            <artifactId>plugin</artifactId>
+                            <version>4.0</version>
+                            <relativePath />
+                          </parent>
+                          <groupId>io.jenkins.plugins</groupId>
+                          <artifactId>empty</artifactId>
+                          <version>1.0.0-SNAPSHOT</version>
+                          <packaging>hpi</packaging>
+                          <name>Empty Plugin</name>
+                          <properties>
+                             <jenkins.version>2.164.3</jenkins.version>
+                          </properties>
+                          <dependencyManagement>
+                            <dependencies>
+                              <dependency>
+                                <groupId>io.jenkins.tools.bom</groupId>
+                                <artifactId>bom-2.164.x</artifactId>
+                                <version>3</version>
+                                <type>pom</type>
+                                <scope>import</scope>
+                              </dependency>
+                            </dependencies>
+                          </dependencyManagement>
+                          <repositories>
+                            <repository>
+                              <id>repo.jenkins-ci.org</id>
+                              <url>https://repo.jenkins-ci.org/public/</url>
+                            </repository>
+                          </repositories>
+                          <pluginRepositories>
+                            <pluginRepository>
+                              <id>repo.jenkins-ci.org</id>
+                              <url>https://repo.jenkins-ci.org/public/</url>
+                            </pluginRepository>
+                          </pluginRepositories>
+                        </project>
+                        """,
+                        """
+                        <?xml version="1.0" encoding="UTF-8"?>
+                        <project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
+                          <modelVersion>4.0.0</modelVersion>
+                          <parent>
+                            <groupId>org.jenkins-ci.plugins</groupId>
+                            <artifactId>plugin</artifactId>
+                            <version>4.0</version>
+                            <relativePath />
+                          </parent>
+                          <groupId>io.jenkins.plugins</groupId>
+                          <artifactId>empty</artifactId>
+                          <version>1.0.0-SNAPSHOT</version>
+                          <packaging>hpi</packaging>
+                          <name>Empty Plugin</name>
+                          <properties>
+                             <jenkins.version>2.164.3</jenkins.version>
+                          </properties>
+                          <dependencyManagement>
+                            <dependencies>
+                              <dependency>
+                                <groupId>io.jenkins.tools.bom</groupId>
+                                <artifactId>bom-2.164.x</artifactId>
+                                <version>10</version>
+                                <type>pom</type>
+                                <scope>import</scope>
+                              </dependency>
+                            </dependencies>
+                          </dependencyManagement>
+                          <repositories>
+                            <repository>
+                              <id>repo.jenkins-ci.org</id>
+                              <url>https://repo.jenkins-ci.org/public/</url>
+                            </repository>
+                          </repositories>
+                          <pluginRepositories>
+                            <pluginRepository>
+                              <id>repo.jenkins-ci.org</id>
+                              <url>https://repo.jenkins-ci.org/public/</url>
+                            </pluginRepository>
+                          </pluginRepositories>
+                        </project>
+                        """));
+    }
+
+    @Test
     void testRemoveDependenciesOverride() {
         rewriteRun(
                 spec -> spec.recipeFromResource(

